File: fire/fire.py
import itertools

rcount, ccount = map(int, raw_input().split())
scene = {}

for r in range(rcount):
    line = raw_input()
    for c, char in zip(range(ccount), line):
        scene[r, c] = char

def vadd(u, v):
    return tuple(map(lambda x, y: x + y, u, v))

def in_bounds(u):
    r, c = u
    return 0 <= r < rcount and 0 <= c < ccount

deltas = [(-1, 0), (0, -1), (0, 1), (1, 0)]

def do():
    joe = set()
    passable = set()
    fire = set()

    for loc in scene.iterkeys():
        thing = scene[loc]
        if thing != '#' and thing != 'F':
            passable.add(loc)
        if thing == 'J':
            joe.add(loc)
        if thing == 'F':
            fire.add(loc)

    for time in itertools.count(1):
        # set things on fire
        newfire = set()
        for here in fire:
            for d in deltas:
                there = vadd(here, d)
                if in_bounds(there):
                    newfire.add(there)
                    passable.discard(there)
        fire.update(newfire)


        # move joe
        newjoe = set()

        for here in joe:
            for d in deltas:
                there = vadd(here, d)
                if not in_bounds(there):
                    # escape !
                    return time
                if there in passable:
                    newjoe.add(there)

        joe = newjoe

        # we're dead
        if not joe:
            return None
        
res = do()
if res is None:
    print 'IMPOSSIBLE'
else:
    print res

File: fire/fire.clj
(ns fire
  (:use
     [clojure.set])
  (:require
     [clojure.contrib.seq :as sq]
     [clojure.contrib.string :as string]))

; utility definitions
(defmacro hash-c [& args]
  `(->> (for ~@args) (apply concat ,,) (apply hash-map ,,)))

(defmacro ffor [& args]
  `(doall (for ~@args)))

(defn parse-line []
  (ffor [s (string/split #"\s+" (read-line))] (read-string s)))

(defn input []
  (let [[height width] (parse-line)
        scene          (vec (ffor [_ (range height)] (vec (read-line))))
        coords         (for [r (range height) c (range width)] [r c])
        passable       (set (filter #(distinct? \# \F (get-in scene %)) coords))
        joe            (set (filter #(= \J (get-in scene %)) coords))
        fire           (set (filter #(= \F (get-in scene %)) coords))]
    [height width passable joe fire]))

(defn border? [rc height width]
  (let [[r c] rc]
    (or (= r 0) (= r (dec height))
        (= c 0) (= c (dec width)))))

(def deltas [[-1 0] [0 -1] [0 1] [1 0]])

(defn adjs [rc]
  (for [d deltas] (map + rc d)))


(defn time-to-escape [height width passable joe fire]
  (loop [t 0
         passable passable 
         joe joe 
         fire fire]
    (cond
      (some #(border? % height width) joe)
        (inc t)
      (empty? joe)
        false
      :else
        (let [newfire (-> (set (mapcat adjs fire))
                        (difference ,, fire) 
                        (intersection ,, passable))
              newjoe  (-> (set (mapcat adjs joe))
                        (difference ,, joe newfire)
                        (intersection ,, passable))
              newpass (difference passable newfire newjoe)]
          (recur (inc t) newpass newjoe newfire)))))

(if-let [t (apply time-to-escape (input))]
  (println t)
  (println "IMPOSSIBLE"))

